### PR TITLE
[SPARK-40029][PYTHON][DOC] Make pyspark.sql.types examples self-contained

### PR DESCRIPTION
**Pull request by Yikun**
### What changes were proposed in this pull request?
This PR proposes to improve the examples in `pyspark.sql.types` by making each example self-contained with a brief explanation and a bit more realistic example.

### Why are the changes needed?
To make the documentation more readable and able to copy and paste directly in PySpark shell.

### Does this PR introduce _any_ user-facing change?
Yes, doc

### How was this patch tested?
Ran each test

### CODE REVIEW

**Comment by Yikun** [comment]
cc @HyukjinKwon

File: python/pyspark/sql/types.py [hunk 1]
```diff
@@ -511,6 +519,14 @@ class MapType(DataType):
 
     Examples
     --------
+    >>> from pyspark.sql.types import IntegerType, FloatType, MapType, StringType
+
+    The below example demonstrates how to create MapType:
```
Comment (HyukjinKwon) [inline python/pyspark/sql/types.py:524]:
```suggestion
    The below example demonstrates how to create class:`MapType`:
```

File: python/pyspark/sql/types.py [hunk 2]
```diff
@@ -511,6 +519,14 @@ class MapType(DataType):
...
+
+    >>> map_type = MapType(StringType(), IntegerType())
+
+    The values of the map can contain null (None) values by default:
```
Comment (HyukjinKwon) [inline python/pyspark/sql/types.py:528]:
```suggestion
    The values of the map can contain null (``None``) values by default:
```

File: python/pyspark/sql/types.py [hunk 3]
```diff
@@ -684,6 +702,30 @@ class StructType(DataType):
...
+    ...     StructField("name", StringType()),
+    ...     StructField("languagesSkills", ArrayType(StringType())),
+    ... ])
+    >>> df = spark.createDataFrame(data=data, schema=schema)
+    >>> df.printSchema()
+    root
+     |-- name: string (nullable = true)
+     |-- languagesSkills: array (nullable = true)
+     |    |-- element: string (containsNull = true)
+    <BLANKLINE>
```
Comment (HyukjinKwon) [inline python/pyspark/sql/types.py:720]: You can remove this blankline and add fix the bottom as below:

```
...
optionflags=doctest.ELLIPSIS | doctest.NORMALIZE_WHITESPACE)
```

File: python/pyspark/sql/types.py [hunk 4]
```diff
@@ -1036,6 +1080,7 @@ def _parse_datatype_string(s: str) -> DataType:
 
     Examples
     --------
+    >>> from pyspark.sql.types import _parse_datatype_string
```
Comment (HyukjinKwon) [inline python/pyspark/sql/types.py:1083]: This is an internal example so let's just leave it.

File: python/pyspark/sql/types.py [hunk 5]
```diff
@@ -1108,6 +1153,8 @@ def _parse_datatype_json_string(json_string: str) -> DataType:
 
     Examples
     --------
+    >>> from pyspark.sql.types import *
+    >>> from pyspark.sql.types import _all_atomic_types, _parse_datatype_json_string
```
Comment (HyukjinKwon) [inline python/pyspark/sql/types.py:1157]: ditto

File: python/pyspark/sql/types.py [hunk 6]
```diff
@@ -1649,6 +1696,8 @@ def _make_type_verifier(
 
     Examples
     --------
+    >>> from pyspark.sql.types import *
+    >>> from pyspark.sql.types import _make_type_verifier
```
Comment (HyukjinKwon) [inline python/pyspark/sql/types.py:1700]: ditto

**Review by HyukjinKwon — APPROVED** [review]
LGTM otherwise

File: python/pyspark/sql/types.py (hunk 3 above)
Comment (Yikun) [reply python/pyspark/sql/types.py]: ah, good suggestion, it was copy from other existing code. Addressed!

and I will also submit a PR to address other places

```
spark git:(type) ✗ grep -r BLANKLINE **/*.py
python/pyspark/ml/stat.py:    <BLANKLINE>
python/pyspark/ml/stat.py:    <BLANKLINE>
python/pyspark/ml/stat.py:    <BLANKLINE>
python/pyspark/ml/stat.py:    <BLANKLINE>
python/pyspark/mllib/tree.py:        <BLANKLINE>
python/pyspark/mllib/tree.py:        <BLANKLINE>
python/pyspark/mllib/tree.py:        <BLANKLINE>
python/pyspark/mllib/tree.py:        <BLANKLINE>
python/pyspark/mllib/tree.py:        <BLANKLINE>
python/pyspark/pandas/frame.py:        <BLANKLINE>
python/pyspark/pandas/series.py:        <BLANKLINE>
python/pyspark/pandas/series.py:        <BLANKLINE>
python/pyspark/pandas/series.py:        <BLANKLINE>
python/pyspark/pandas/series.py:        <BLANKLINE>
python/pyspark/sql/dataframe.py:        <BLANKLINE>
```

File: python/pyspark/sql/types.py [hunk 7]
```diff
@@ -2137,7 +2181,9 @@ def _test() -> None:
     sc = SparkContext("local[4]", "PythonTest")
     globs["sc"] = sc
```
Comment (dcoliversun) [inline python/pyspark/sql/types.py:2182]: `sc` is redundant? I don't find where use it. If it's redundant, better to remove it.

Comment (Yikun) [reply python/pyspark/sql/types.py:2182]: Good catch! thanks!

File: python/pyspark/sql/types.py [hunk 8]
```diff
@@ -684,6 +702,28 @@ class StructType(DataType):
     ...     StructField("f2", IntegerType(), False)])
     >>> struct1 == struct2
     False
+
+    The below example demonstrates how to create a struct using class:`StructType`
+    and class:`StructField` on DataFrame:
```
Comment (xinrong-meng) [inline python/pyspark/sql/types.py:707]: nit: `create a struct ... on DataFrame` may not be very clear, how about `create a DataFrame based on a struct created using ...`?

**Review by xinrong-meng — APPROVED** [review]

File: python/pyspark/sql/types.py (hunk 8 above)
Comment (Yikun) [reply python/pyspark/sql/types.py]: Thanks, addressed.

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**